src/ripple/basics/Buffer.h

```diff
@@ -113,118 +153,150 @@ class Buffer
     {
         // Ensure the slice isn't a subset of the buffer.
         assert(
-            s.size() == 0 || size_ == 0 || s.data() < p_.get() ||
-            s.data() >= p_.get() + size_);
+            s.size() == 0 || size_ == 0 || s.data() < data_ ||
+            s.data() >= data_ + size_);
+
+        std::copy_n(s.data(), s.size(), alloc(s.size()));
 
-        if (auto p = alloc(s.size()))
-            std::memcpy(p, s.data(), s.size());
         return *this;
     }
 
     /** Returns the number of bytes in the buffer. */
-    std::size_t
+    [[nodiscard]] std::size_t
     size() const noexcept
     {
         return size_;
     }
 
-    bool
+    [[nodiscard]] bool
     empty() const noexcept
     {
         return 0 == size_;
     }
 
     operator Slice() const noexcept
     {
-        if (!size_)
-            return Slice{};
-        return Slice{p_.get(), size_};
+        return Slice{data_, size_};
     }
 
-    /** Return a pointer to beginning of the storage.
+    /** Return a pointer to the storage.
         @note The return type is guaranteed to be a pointer
               to a single byte, to facilitate pointer arithmetic.
     */
     /** @{ */
-    std::uint8_t const*
+    [[nodiscard]] std::uint8_t const*
```

Comment:
how does one know the size of the valid data?
the `buf_` has 112 bytes of storage. If I'm using 30 bytes, the remaining will be filled with garbage data. Are the consumers of this function expected to check the size too?

Reply:
Users of this class have always been expected the know the amount of data they've stored into the buffer. Whether that's a good idea or not, or whether that functionality could be consolidated into this class is another question.

Reply:
okay